cpp/src/arrow/array.h

```diff
@@ -801,6 +782,38 @@ ARROW_EXTERN_TEMPLATE NumericArray<TimestampType>;
 ARROW_EXPORT
 Status ValidateArray(const Array& array);
 
+/// \brief Perform validation check to determine if all indices are within
+/// valid range (0 <= index < upper_bound)
+///
+/// \param[in] indices array of indices
+/// \param[in] upper_bound upper bound of valid range for indices
+/// \return Status
+template <typename ArrowType>
+Status ValidateArray(const std::shared_ptr<Array>& indices, const int64_t upper_bound) {
```

Comment:
Can you move this to `array.cc` and rename to `ValidateDictionaryIndices`? This template does not need to be visible in the public API